### PR TITLE
feat(tun): windows auto-route

### PR DESCRIPTION
**Pull request by ibigbug**
### 🤔 This is a ...

- [ ] New feature
- [ ] Bug fix
- [ ] Performance optimization
- [x] Enhancement feature
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 🔗 Related issue link

#559 

### 💡 Background and solution

https://learn.microsoft.com/en-us/windows/win32/rras/add-and-update-routes-using-rtmaddroutetodest

### 📝 Changelog

```
tun:
  enable: true
  device-id: "dev://utun1989"
  route-all: false
  routes:
    - 0.0.0.0/1
    - 128.0.0.0/1
```

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [ ] Doc is updated/provided or not needed
- [ ] Changelog is provided or not needed

### CODE REVIEW

File: clash_lib/src/proxy/tun/inbound.rs [hunk 1]
```diff
@@ -177,14 +186,56 @@ pub fn get_runner(
...
+                "route_all is enabled, all traffic will be routed through the tun \
+                 interface"
+            );
+            let default_routes = vec![
+                IpNet::new(std::net::IpAddr::V4(Ipv4Addr::new(0, 0, 0, 0)), 1)
+                    .unwrap(),
+                IpNet::new(std::net::IpAddr::V4(Ipv4Addr::new(128, 0, 0, 0)), 1)
+                    .unwrap(),
+            ];
+            for r in default_routes {
```
Comment (VendettaReborn) [inline clash_lib/src/proxy/tun/inbound.rs]: shall we make this a two step operation of build_route**s** & add_route**s**, thus is can be separated into the `routes` module

File: clash_lib/src/proxy/tun/routes/windows.rs [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+        error!("failed to add route: {}", err);
+        Err(new_io_error(err.to_string().as_str()))
+    }
+}
+
+/// Add a route to the routing table.
+/// https://learn.microsoft.com/en-us/windows/win32/rras/add-and-update-routes-using-rtmaddroutetodest
+/// FIXME: figure out why this doesn't work https://stackoverflow.com/questions/43632619/how-to-properly-use-rtmv2-and-rtmaddroutetodest
+#[allow(dead_code)]
+pub fn add_route_that_does_not_work(
```
Comment (VendettaReborn) [inline clash_lib/src/proxy/tun/routes/windows.rs:55]: this is how `net-route` handled the routing table: https://github.com/johnyburd/net-route/blob/f83a447bd13404225c8a78db71af208ee8e7885c/src/platform_impl/windows.rs#L178. it seems that they took a different approach

**Comment by Itsusinn** [comment]
实测能通吗？windows api那部分实在是看不懂

**Comment by Itsusinn** [comment]
别修了吧，感觉没啥用，微软的文档都不全

**Review by Itsusinn — APPROVED** [review]

**Review by Itsusinn — APPROVED** [review]

**Review by VendettaReborn — APPROVED** [review]

File: clash_lib/src/proxy/tun/routes/windows.rs (hunk 2 above)
Comment (ibigbug) [reply clash_lib/src/proxy/tun/routes/windows.rs:55]: yeah i checked their code they used a different set of API. which makes sense. maybe I'll switch to using the crate for this too when I do the macos one

File: clash_lib/src/proxy/tun/inbound.rs (hunk 1 above)
Comment (ibigbug) [reply clash_lib/src/proxy/tun/inbound.rs]: i'm hoping the 0/1 and 128/1 can work on all platforms so the build_route will just be the same 2 entries

**Pull request merged**